### ISSUE TITLE
Error configuring veth under k8s

### ISSUE THREAD

**Issue by caseydavenport** [projectcalico/calico-cni#135]
Seen with Golang CNI plugin when starting ~10 pods at once on a host.

```
Aug 09 23:29:32 k8s-node-01 kubelet[1954]: E0809 23:29:32.862003   29985 k8s.go:113] Error configuring networking for default.bb-4248473792-vz4le: failed to rename veth: veth81101f4c to cali6e3a56ba1ef (device or resource busy)
```

Because of #128, the IP allocation isn't cleaned up even when the pod is re-scheduled and successfully started a few seconds later.

**Comment by tomdee** [comment]
@caseydavenport Any ideas what we should do to fix this? Add a retry mechanism? Or is it a live-with once #128 is fixed?

**Comment by caseydavenport** [comment]
I find it a bit fishy that it occurs at all.  I think this is a live-with now that #128 is fixed until we find out that it causes some other problem.

**Comment by tomdee** [comment]
I'm going to keep reopen this as I'm hitting it too. TBD whether it's a ship stopper.

```
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: time="2016-08-23T19:34:04Z" level=info msg="Extracted identifiers" Hostname=172.17.4.201 OrchestratorID=k8s WorkloadID=default.busybox27504
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: Calico CNI checking for existing endpoint: <nil>
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: time="2016-08-23T19:34:04Z" level=info msg="Extracted identifiers" Hostname=172.17.4.201 OrchestratorID=k8s WorkloadID=default.busybox27504
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: time="2016-08-23T19:34:04Z" level=info msg="Populated endpoint" WorkloadID=default.busybox27504 endpoint=&{{workloadEndpoint v1} {{} eth0 default.busybox27504 k8s 172.17
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: I0823 19:34:04.187685   11483 operation_executor.go:720] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/default-token-mwqa5" (spec.Name: "default-token-mwq
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: time="2016-08-23T19:34:04Z" level=info msg="Fetched K8s labels" WorkloadID=default.busybox27504 labels=map[calico/k8s_ns:default]
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: Calico CNI using IPs: [{{10.2.73.3 ffffffff}}]
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: time="2016-08-23T19:34:04Z" level=info msg="Cleaning up IP allocations for failed ADD"
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: E0823 19:34:04.236497   11483 cni.go:166] Error adding network: failed to rename veth: veth7652dfb5 to cali1dab2940092 (device or resource busy)
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: E0823 19:34:04.236855   11483 cni.go:123] Error while adding to cni network: failed to rename veth: veth7652dfb5 to cali1dab2940092 (device or resource busy)

```

**Comment by tomdee** [comment]
```Aug 23 19:34:04 w1 kubelet-wrapper[11483]: time="2016-08-23T19:34:04Z" level=info msg="Fetched K8s labels" WorkloadID=default.busybox27504 labels=map[calico/k8s_ns:default]
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: Calico CNI using IPs: [{{10.2.73.3 ffffffff}}]
Aug 23 19:34:04 w1 kernel: IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
Aug 23 19:34:04 w1 rkt[1886]: 2016-08-23 19:34:04,201 [INFO][2016/22] calico.felix.splitter 129: Interface veth7652dfb5 state changed
Aug 23 19:34:04 w1 rkt[1886]: 2016-08-23 19:34:04,202 [INFO][2016/22] calico.felix.splitter 129: Interface veth7652dfb5 state changed
Aug 23 19:34:04 w1 systemd-networkd[1389]: veth7652dfb5: Gained carrier
Aug 23 19:34:04 w1 systemd-timesyncd[670]: Network configuration changed, trying to establish connection.
Aug 23 19:34:04 w1 rkt[1886]: 2016-08-23 19:34:04,206 [INFO][2016/22] calico.felix.splitter 129: Interface veth7652dfb5 state changed
Aug 23 19:34:04 w1 rkt[1886]: 2016-08-23 19:34:04,206 [INFO][2016/22] calico.felix.splitter 129: Interface veth7652dfb5 state changed
Aug 23 19:34:04 w1 systemd-networkd[1389]: veth7652dfb5: Lost carrier
Aug 23 19:34:04 w1 kernel: IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
Aug 23 19:34:04 w1 systemd-networkd[1389]: veth7652dfb5: Removing non-existent address: fe80::5455:73ff:fe6f:6c57/64 (valid forever)
Aug 23 19:34:04 w1 systemd-udevd[12051]: Could not generate persistent MAC address for veth7652dfb5: No such file or directory
Aug 23 19:34:04 w1 rkt[1886]: 2016-08-23 19:34:04,221 [INFO][2016/22] calico.felix.splitter 129: Interface veth7652dfb5 state changed
Aug 23 19:34:04 w1 systemd-networkd[1389]: veth7652dfb5: Gained carrier
Aug 23 19:34:04 w1 systemd-timesyncd[670]: Synchronized to time server 65.182.224.60:123 (0.coreos.pool.ntp.org).
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: time="2016-08-23T19:34:04Z" level=info msg="Cleaning up IP allocations for failed ADD"
Aug 23 19:34:04 w1 kubelet-wrapper[11483]: E0823 19:34:04.236497   11483 cni.go:166] Error adding network: failed to rename veth: veth7652dfb5 to cali1dab2940092 (device or resource busy)
```

**Comment by tomdee** [comment]
```	if err := netlink.LinkSetDown(veth); err != nil {
		return fmt.Errorf("failed to set %q DOWN: %v", vethName, err)
	}

	if err := netlink.LinkSetName(veth, newVethName); err != nil {
		return fmt.Errorf("failed to rename veth: %v to %v (%v)", vethName, newVethName, err)
	}

	if err := netlink.LinkSetUp(veth); err != nil {
		return fmt.Errorf("failed to set %q UP: %v", vethName, err)
	}```

**Comment by tomdee** [comment]
@caseydavenport I put up a PR with a fix (/retry mechanism) - Could you try to repro the issue with this binary? https://506-39909298-gh.circle-artifacts.com/0//home/ubuntu/calico-cni/dist/calico

**Comment by tomdee** [comment]
Fixed in v1.4.1
